### ISSUE TITLE
Pagination for post listing?

### ISSUE THREAD

**Issue by MilesMcBain** [rstudio/distill#162]
This is mostly blog thing.

Over time, as you accumulate posts the listing can get quite long and it begins to affect the homepage load time, especially if there are cover images for each post. Image optimisation can mitigate it to a degree, but it's really an uphill battle against something that will inevitably become a problem the more you post.

Pagination would be a nice to have to deal with this. I know it is possible to say exactly which posts appear on the page but it would be good to be able to say 'the latest n' where n is a number of my choosing. Combine that with an archive listing perhaps?.  

Another option would be full pagination with n posts per page divided over as many necessary pages as necessary.

The tag filtering might get a bit tricky.

**Comment by maelle** [comment]
A Hugo blog but it reminded me of this issue

https://twitter.com/StatnMap/status/1305448768947585026

**Comment by jjallaire** [comment]
Here is what we currently do to mitigate home page load time performance problems:

1) Metadata for all posts is included on the main listing page, however images are not loaded by default for any post listings.

2) After initial loading, we incrementally load the images for the first 25 articles, and mark the rest of the articles invisible.

3) There is a "More articles" link at the bottom of the first 25 articles. If you click that it will reveal the rest of the articles (and then incrementally load their images).

The idea is that readers will rarely want to scroll past the first 25 items (rather they'd use category browsing or search once that is implemented), _and_ that loading 25 images incrementally (i.e. not blocking initial page display/layout) is reasonable.

This is intended as a good enough approximation of a full pagination solution, which as @MilesMcBain points out would get tricky with category/tag filtering given that the site is static. I've seen other static site generators do fully static per-category paginated views but that does impact page/site build time and would also take a bit more time to implement than I have right now.

What do people think about the current behavior and how it may or may not need to change? To me the 25 articles seems reasonable given the images are loaded incrementally (so you see the images for the articles "above the fold" come in right away and then the rest load offscreen). It also seems like search would be preferable to pagination as a way to locate a needle in a haystack of posts.

**Comment by MilesMcBain** [comment]
Thanks for the clarification JJ. This actually sounds really good. Some control of the magic number 25 might be nice but it's not the end of the world.

So I think what I thought was my page loading slowly is actually the sequential loading effect. The only thing is on my page the images do not appear to load in top down order (LIFO). The first one loads, then a couple in the middle then the second one.

Size is almost certainly a factor but this implies some kind of batching going on with the sequential loading? Maybe this could be tweaked a bit to prioritise the images that are likely to appear on screen? e.g. 1 - 3. 

If you want to see for yourself, it's https://milesmcbain.xyz.

**Comment by jjallaire** [comment]
For loading I'm just queuing up requests for all of the images. The browser will download these in parallel according to whatever scheme it has for managing concurrent requests, etc. This likely is the cause of the non-determinism in order you are seeing. I could strictly control the order by chaining the downloads, but I'm afraid that this would end up being much slower overall due to the loss of parallelism.

So here's a compromise: I won't do deferred loading for the first 10 articles on the page (https://github.com/rstudio/distill/commit/976c7373886d1c525a0e9194d25e1206f430e0d5). This should cause them to render at a much higher priority level and also get processed in order. If you think e.g. 3 or 5 or 8 images is a better number let me know. I could certainly make it configurable but there is a high likelihood that no one will ever bother to configure this so I'd rather just pick the right default.

**Comment by MilesMcBain** [comment]
Thanks @jjallaire! I tested out the new changes. I like them. I suppose it's a matter of taste, but the site feels faster to me. I spend less time looking at a white screen. 

Apart from the apparent speed increase, the main difference is now in my browser (Firefox) is that you can see some images "peel" into existence rather than "pop". I don't really mind this. At least it happens quickly. 

You could probably set the number of images lower than 10. 5 would also be fine I reckon, on my screen I only see 3 @ 100% zoom.

**Comment by jjallaire** [comment]
Glad to hear it works better now! Agreed that 10 images is too many (changed it to 5).
